File: 10-multi-agent-demo/go/orchestrator.go
// orchestrator.go —— 多 agent 编排：顺序 / 并行。整文件 cp 进项目即可。
package main

import (
	"context"
	"strings"
	"sync"
)

type Step struct {
	ID        string
	Agent     string
	Task      string
	DependsOn []string
}

const maxCtxChars = 400

func buildContext(results map[string]string, deps []string) string {
	parts := []string{}
	for _, d := range deps {
		v, ok := results[d]
		if !ok {
			continue
		}
		if len(v) > maxCtxChars {
			v = v[:maxCtxChars] + "...(已截断)"
		}
		parts = append(parts, "["+d+"]\n"+v)
	}
	return strings.Join(parts, "\n\n")
}

// RunSequential 按顺序跑；每步可以拿到 DependsOn step 的输出作 context。
func RunSequential(ctx context.Context, agents map[string]*Agent, workflow []Step) (map[string]string, error) {
	results := map[string]string{}
	for _, s := range workflow {
		out, err := agents[s.Agent].Execute(ctx, s.Task, buildContext(results, s.DependsOn))
		if err != nil {
			return results, err
		}
		results[s.ID] = out
	}
	return results, nil
}

// RunParallel 并发跑（彼此独立，不解 DAG）。
func RunParallel(ctx context.Context, agents map[string]*Agent, steps []Step) map[string]string {
	results := map[string]string{}
	var mu sync.Mutex
	var wg sync.WaitGroup
	for _, s := range steps {
		s := s
		wg.Add(1)
		go func() {
			defer wg.Done()
			out, err := agents[s.Agent].Execute(ctx, s.Task, "")
			if err != nil {
				out = "ERROR: " + err.Error()
			}
			mu.Lock()
			results[s.ID] = out
			mu.Unlock()
		}()
	}
	wg.Wait()
	return results
}

File: 10-multi-agent-demo/python/orchestrator.py
"""orchestrator.py —— 多 agent 编排：顺序 / 并行 / 带依赖的 DAG。
整文件 cp 进项目即可。

每个 Step：用哪个 agent 跑、跑什么 task、依赖哪些前置 step 的输出作为 context。
依赖结果会按 step.id 拼成 "id: 内容" 喂给被依赖的 step。
"""
from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field

from agent import Agent


@dataclass
class Step:
    id: str
    agent: str
    task: str
    depends_on: list[str] = field(default_factory=list)


def _context(results: dict[str, str], deps: list[str], max_chars: int = 400) -> str:
    """把依赖 step 的输出拼成 context；过长就截断（防 LLM context 爆）。"""
    parts = []
    for d in deps:
        if d in results:
            content = results[d]
            if len(content) > max_chars:
                content = content[:max_chars] + "...(已截断)"
            parts.append(f"[{d}]\n{content}")
    return "\n\n".join(parts)


def run_sequential(agents: dict[str, Agent], workflow: list[Step]) -> dict[str, str]:
    """顺序执行；每个 step 拿到之前依赖 step 的产物作为 context。"""
    results: dict[str, str] = {}
    for step in workflow:
        agent = agents[step.agent]
        ctx = _context(results, step.depends_on)
        results[step.id] = agent.execute(step.task, ctx)
    return results


def run_parallel(agents: dict[str, Agent], steps: list[Step]) -> dict[str, str]:
    """并行执行（彼此独立的 step）。不解依赖 —— 调用方保证 steps 之间没 depends_on。"""
    results: dict[str, str] = {}
    with ThreadPoolExecutor(max_workers=len(steps)) as ex:
        futs = {ex.submit(agents[s.agent].execute, s.task): s for s in steps}
        for fut, s in futs.items():
            results[s.id] = fut.result()
    return results
